src/persist-client/src/internal/apply.rs

```diff
@@ -44,12 +155,16 @@ pub struct Applier<K, V, T, D> {
     pub(crate) shard_metrics: Arc<ShardMetrics>,
     pub(crate) state_versions: Arc<StateVersions>,
 
-    // TODO: Wrap this in an Arc<tokio::sync::Mutex<_>> and have a single shared
-    // one for each shard in a process.
-    //
-    // NB: This is very intentionally not pub(crate) so that it's easy to reason
-    // very locally about the duration of Mutex holds.
-    state: TypedState<K, V, T, D>,
+    // Access to the shard's state, shared across all handles created by the same
+    // PersistClientCache. The state is wrapped in a std::sync::Mutex, disallowing
+    // access across await points. Access should be always be kept brief, and it
+    // is expected that other handles may advance the state at any time this Applier
+    // is not holding the lock.
+    state: Arc<Mutex<TypedState<K, V, T, D>>>,
```

Comment:
this should be a RwLock, no? definitely feels like theres some concurrency upside for readers

I checked a couple weeks ago and both linux and macos treat queued writers as a barrier (to prevent writer starvation) which is good but means grabbing two read locks could deadlock. we shouldn't ever need to do that, but we might want to play with code structure to make it easier to reason about it not being an issue

Reply:
👍 I'll rejigger this with a RwLock next. I wasn't sure what the right trade off is since RwLocks have a bit more baseline overhead than Mutexes, but as I've poked more, some read operations like encoding state into protobuf scare me a bit, especially in the pathological cases

Reply:
Tempted to wire in a config knob to toggle this at startup, though I might do that after this PR (assuming baseline perf is fine) is merged to avoid merge conflicts with largely unrelated code.